presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsTracker.java

```diff
@@ -176,21 +194,69 @@ public Map<PlanNodeWithHash, PlanStatisticsWithSourceInfo> getQueryStats(QueryIn
                                 joinNodeStatistics,
-                                tableWriterNodeStatistics);
-                        if (planStatistics.containsKey(planNodeWithHash)) {
-                            newPlanNodeStats = planStatistics.get(planNodeWithHash).getPlanStatistics().update(newPlanNodeStats);
+                                tableWriterNodeStatistics,
+                                partialAggregationStatistics);
+                        if (planStatisticsMap.containsKey(planNodeWithHash)) {
+                            newPlanNodeStats = planStatisticsMap.get(planNodeWithHash).getPlanStatistics().update(newPlanNodeStats);
+                        }
+                        PlanStatisticsWithSourceInfo planStatsWithSourceInfo = new PlanStatisticsWithSourceInfo(
+                                planNode.getId(),
+                                newPlanNodeStats,
+                                new HistoryBasedSourceInfo(Optional.of(hash), Optional.of(inputTableStatistics)));
+                        planStatisticsMap.put(planNodeWithHash, planStatsWithSourceInfo);
+
+                        if (isAggregation(planNode, AggregationNode.Step.FINAL) && ((AggregationNode) planNode).getAggregationId().isPresent()) {
+                            // we're doing a depth-first traversal of the plan tree: cache the final agg so that when we encounter the partial agg we can come back
+                            // and update the partial agg statistics
+                            aggregationNodeMap.put(((AggregationNode) planNode).getAggregationId().get(), new PartialAggregationStatsInfo(planNodeWithHash, planStatsWithSourceInfo));
                         }
-                        planStatistics.put(
-                                planNodeWithHash,
-                                new PlanStatisticsWithSourceInfo(
-                                        planNode.getId(),
-                                        newPlanNodeStats,
-                                        new HistoryBasedSourceInfo(Optional.of(hash), Optional.of(inputTableStatistics))));
                     }
                 }
             }
         }
-        return ImmutableMap.copyOf(planStatistics);
+        return ImmutableMap.copyOf(planStatisticsMap);
+    }
+
+    private static void updatePartialAggregationStatistics(
+            AggregationNode partialAggregationNode,
+            Map<Integer, PartialAggregationStatsInfo> aggregationNodeStats,
+            PartialAggregationStatistics partialAggregationStatistics,
+            Map<PlanNodeWithHash, PlanStatisticsWithSourceInfo> planStatisticsMap)
+    {
+        if (!partialAggregationNode.getAggregationId().isPresent() || !aggregationNodeStats.containsKey(partialAggregationNode.getAggregationId().get())) {
+            return;
+        }
+
+        PartialAggregationStatsInfo partialAggregationStatsInfo = aggregationNodeStats.get(partialAggregationNode.getAggregationId().get());
```

Comment:
we cache historical partial stats execution at the level of the final aggregation because the partial agg is not part of the canonical plan

Reply:
aggregationNodeStats is a helper structure where we accumulate information about results of aggregation execution: the statistics comes from the partial agg node but ends up being tracked in the final agg node because the partial agg is not part of the canonical plan

Reply:
oops, I misread your comment @feilong-liu 
you are right, the variable should be named finalAggregationStatsInfo, will fix that in the code